File: extfs/models/node_item.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type NodeItemFields struct {
	Name     string `form:"name" binding:"required" json:"name"`
	FilePath string `form:"filePath" binding:"required" json:"filePath"`
	Enabled  *bool  `form:"enabled" binding:"required" json:"enabled"`
}

type NodeItem struct {
	ID                 uint           `gorm:"primarykey" json:"id" form:"id"`
	CreatedAt          time.Time      `json:"createdAt" form:"createdAt"`
	UpdatedAt          time.Time      `json:"updatedAt" form:"updatedAt"`
	DeletedAt          gorm.DeletedAt `gorm:"index" json:"deletedAt" form:"deletedAt"`
	Name               string         `gorm:"size:255;index" json:"name" form:"name"`
	FilePath           string         `json:"filePath" form:"filePath"`
	FileType           string         `gorm:"size:1;index;" json:"fileType" form:"fileType"`
	Enabled            *bool          `gorm:"index" json:"enabled" form:"enabled"`
	Available          bool           `gorm:"-:all" json:"available" form:"available"`
	Size               int64          `gorm:"-:all" json:"size" form:"size"`
	TagQuantity        uint           `gorm:"-:all" json:"tagQuantity" form:"tagQuantity"`
	PendingTagQuantity uint           `gorm:"-:all" json:"pendingTagQuantity" form:"pendingTagQuantity"`
}

File: web/src/api/node_item.ts
import { withJSONBody, withMethod, withPath } from "fetch-utils";
import { fetchMany, fetchOne } from "./base";

export interface ExtFSNodeItemAPI {
  selectAllExtFSNodeItems(): Promise<ExtFSNodeItem[]>;
  saveExtFSNodeItem(
    fields: ExtFSNodeItemFields,
    id?: ExtFSNodeItem["id"]
  ): Promise<ExtFSNodeItem>;
  selectExtFSNodeItem(id: ExtFSNodeItem["id"]): Promise<ExtFSNodeItem>;
  deleteExtFSNodeItem(id: ExtFSNodeItem["id"]): Promise<void>;
}

export type ExtFSNodeItem = {
  id: number;
  name: string;
  filePath: string;
  fileType: "F" | "D";
  mimeType: string;
  size: number;
  enabled: boolean;
  available: boolean;
  createdAt: string;
  updatedAt: string;
  deletedAt: string;
};

export type ExtFSNodeItemFields = Omit<
  ExtFSNodeItem,
  | "id"
  | "createdAt"
  | "updatedAt"
  | "deletedAt"
  | "available"
  | "fileType"
  | "mimeType"
  | "size"
  | "tagQuantity"
  | "pendingTagQuantity"
>;

export async function selectAllExtFSNodeItems(): Promise<ExtFSNodeItem[]> {
  const [_, nodeItems] = await fetchMany(withPath("extfs/node-items", "merge"));
  return nodeItems;
}

export async function saveExtFSNodeItem(
  fields: ExtFSNodeItemFields,
  id?: ExtFSNodeItem["id"]
): Promise<ExtFSNodeItem> {
  return await fetchOne(
    withPath(`extfs/node-items${id ? `/${id}` : ""}`, "merge"),
    withMethod(id ? "PATCH" : "POST"),
    withJSONBody(fields)
  );
}

export async function selectExtFSNodeItem(
  id: ExtFSNodeItem["id"]
): Promise<ExtFSNodeItem> {
  return await fetchOne(withPath(`extfs/node-items/${id}`, "merge"));
}

export async function deleteExtFSNodeItem(id: ExtFSNodeItem["id"]) {
  return await fetchOne(
    withPath(`extfs/node-items/${id}`, "merge"),
    withMethod("DELETE")
  );
}
